### PR TITLE
refactor: build themes with whiskers

### PR DESCRIPTION
**Pull request by dacid44**
I made a whiskers template file for COSMIC, and thought it'd be good to submit it to this repository. If you think this is a good idea, I will probably make another one for cosmic-term as well, and can write some documentation on how to use it. Having a template file makes it easier for derivative themes (such as [carburetor](https://github.com/ozwaldorf/carburetor)) to make use of Catppuccin.

I tried to keep most of the same options as the `generate.py` has, accessible as frontmatter overrides. I may still revise how the filenames are generated to match those already in the repo.

(marked as draft until I finish the cosmic-term template)

### CODE REVIEW

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
{%- set palette = flavor.colors -%}
```

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
    palette: {{ if(cond=flavor.dark, t="Dark", f="Light") }}((
```

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
    {% if not flavor.dark -%}
```

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
        {{ self::palette_color(name="blue", color=blue) }}
        {{ self::palette_color(name="red", color=red) }}
        {{ self::palette_color(name="green", color=green) }}
        {{ self::palette_color(name="yellow", color=yellow) }}
        {{ self::palette_color(name="gray_1", color=mantle) }}
        {{ self::palette_color(name="gray_2", color=base) }}
        {{ self::palette_color(name="gray_3", color=surface0) }}
        {{ self::palette_color(name="neutral_0", color=crust) }}
        {{ self::palette_color(name="neutral_1", color=mantle) }}
        {{ self::palette_color(name="neutral_2", color=base) }}
        {{ self::palette_color(name="neutral_3", color=surface0) }}
        {{ self::palette_color(name="neutral_4", color=surface1) }}
        {{ self::palette_color(name="neutral_5", color=surface2) }}
        {{ self::palette_color(name="neutral_6", color=overlay0) }}
        {{ self::palette_color(name="neutral_7", color=overlay1) }}
        {{ self::palette_color(name="neutral_8", color=overlay2) }}
        {{ self::palette_color(name="neutral_9", color=subtext0) }}
        {{ self::palette_color(name="neutral_10", color=subtext1) }}
        {{ self::palette_color(name="bright_green", color=green) }}
        {{ self::palette_color(name="bright_red", color=red) }}
        {{ self::palette_color(name="bright_orange", color=peach) }}
        {{ self::palette_color(name="accent_blue", color=blue) }}
        {{ self::palette_color(name="accent_red", color=red) }}
        {{ self::palette_color(name="accent_green", color=green) }}
        {{ self::palette_color(name="accent_warm_grey", color=overlay2) }}
        {{ self::palette_color(name="accent_orange", color=peach) }}
        {{ self::palette_color(name="accent_yellow", color=yellow) }}
        {{ self::palette_color(name="accent_purple", color=lavender) }}
        {{ self::palette_color(name="accent_pink", color=pink) }}
        {{ self::palette_color(name="accent_indigo", color=mauve) }}
        {{ self::palette_color(name="ext_warm_grey", color=overlay2) }}
        {{ self::palette_color(name="ext_orange", color=peach) }}
        {{ self::palette_color(name="ext_yellow", color=yellow) }}
        {{ self::palette_color(name="ext_blue", color=blue) }}
        {{ self::palette_color(name="ext_purple", color=lavender) }}
        {{ self::palette_color(name="ext_pink", color=pink) }}
        {{ self::palette_color(name="ext_indigo", color=mauve) }}
```

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
    {{ self::other_color(name="bg_color", color=base, alpha=bg_alpha) }}
    {{ self::other_color(name="text_tint", color=text) }}
    {{ self::other_color(name="accent", color=palette[accent]) }}
    {{ self::other_color(name="success", color=green) }}
    {{ self::other_color(name="warning", color=yellow) }}
    {{ self::other_color(name="destructive", color=red) }}
    {{ self::other_color(name="window_hint", color=palette[window_hint]) }}
```

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
    {{ self::other_color(name="neutral_tint", color=overlay1) }}
    {{ self::other_color(name="primary_container_bg", color=surface0) }}
    {{ self::other_color(name="secondary_container_bg", color=surface1) }}
```

**Review by uncenter — CHANGES_REQUESTED** [review]

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]:
```suggestion
window_hint_color: ""
bg_alpha: 1.0
frosted: false
outer_gap_size: 0
inner_gap_size: 8
active_hint_size: 3
whiskers:
    version: 2.5.1
    matrix:
    - variant: ["round", "slightlyround", "square"]
    - flavor
    - accent
    filename: "themes/Catppuccin{{ variant | capitalize }}-{{ flavor.identifier | capitalize }}-{{ accent | capitalize }}.ron"
```

Comment (dacid44) [reply templates/cosmic-settings.tera]: I added the name field to make it easy for derivative themes to customize the name. Should it be patched in a derivative theme instead?

**Comment by EdenQwQ** [comment]
This looks good to me! I didn't know whiskers before and it seems like a good tool for allowing modifications of the color palette. Thanks for writing the template file. I'll merge this once you finish writing the template for cosmic-term and the documentations.

**Comment by dacid44** [comment]
Yeah, I've been using it because while the bluish tint of Catppuccin isn't to my personal preference, I like the theming system so I've been using [Carburetor](https://github.com/ozwaldorf/carburetor) a fair bit, which has more neutral tones with bolder accents. Having whiskers templates makes color overrides super easy.

**Comment by dacid44** [comment]
OK, I've just added the template for `cosmic-term`. I'll probably work on the documentation tomorrow.

**Comment by dacid44** [comment]
How should the naming work for the main theme with the different roundness variants? Currently it names it something like `Catppuccin-Slightlyround-Frappe-Peach.ron`, and generates a file for all three roundness variants. Should I instead have it default to round, and allow it to be configurable through frontmatter variables if a user wants a different roundness (and not include the roundness in the filename?)

**Comment by EdenQwQ** [comment]
> How should the naming work for the main theme with the different roundness variants? Currently it names it something like `Catppuccin-Slightlyround-Frappe-Peach.ron`, and generates a file for all three roundness variants. Should I instead have it default to round, and allow it to be configurable through frontmatter variables if a user wants a different roundness (and not include the roundness in the filename?)

I personally prefer to have a default roundness, because this repository focus on the theme color but not something like roundness (which can be directly customized through cosmic settings).

**Comment by sgoudham** [comment]
We should strive to only modify the colours and follow upstream defaults where possible. 

This PR should also ideally remove all the python code and update the README to give an example on how to override the defaults that we've set to different values.

File: templates/cosmic-settings.tera
Comment (uncenter) [reply templates/cosmic-settings.tera]: As this is a template for a Catppuccin theme I don't see why the name shouldn't be Catppuccin, that's what the colors are anyway.

Comment (uncenter) [reply templates/cosmic-settings.tera]: But I guess seeing as you are using a different color scheme with it.

File: templates/cosmic-settings.tera
Comment (uncenter) [inline templates/cosmic-settings.tera]: I would also rename these macros, the current names here are very vague.

File: templates/cosmic-settings.tera
Comment (sgoudham) [reply templates/cosmic-settings.tera]: > I added the name field to make it easy for derivative themes to customize the name. Should it be patched in a derivative theme instead? - @dacid44 

I understand where you're coming from but I think this sets a bit of a weird precedent for the organisation where we allow our entire brand to be overridable. I'm all for customisation and it's obviously designed into whiskers with the way our colour overrides flag / frontmatter works but it feels odd to have that be overridable by default _within the Catppuccin organisation_.

Furthermore, I don't believe it makes any sense to only override "Catppuccin". If you were making an override for "MyAwesomeTheme" then your whiskers override would just result in "MyAwesomeTheme-Latte", "MyAwesomeTheme-Frappe", etc. It's also worth mentioning that including "Catppuccin" in the filename might not be required since there is a `name` field in the file itself which I'd imagine COSMIC uses as the display name on the UI.

With all of the above in mind, I'd push back on accepting the template with the frontmatter override and would either accept "Catppuccin" hardcoded into the filename or potentially not included in the first place. Hope that makes sense 👍

File: templates/cosmic-settings.tera
Comment (dacid44) [reply templates/cosmic-settings.tera]: What would a better name for them be?

File: templates/cosmic-settings.tera
Comment (dacid44) [reply templates/cosmic-settings.tera]: That's fair, there would need to be a bit of extra work downstream anyway. I'll go ahead and take out the name parameter.

File: templates/cosmic-settings.tera
Comment (uncenter) [reply templates/cosmic-settings.tera]: You tell me! I don't know what the underlying difference is between these two formats. Seeing as it is a Ron file, and one of them is wrapped in Some(), maybe have the second one be `optional_color`? There is also the difference of the second one including an alpha value. Can both have an alpha value or is it just the optional color?

Comment (dacid44) [reply templates/cosmic-settings.tera]: Well, I wrote one of them to fill in the colors in the palette section, which are always alpha 1.0, and are not optional, and the other fills in various other colors, all of which are optional (but this macro fills them in as `Some`) and one of which has an alpha value which can be changed. I couldn't think of a more descriptive name for these, so I used the name from the `generate.py`, which simply lists these colors in [`other_map`](https://github.com/catppuccin/cosmic-desktop/blob/11231f34a5de36e09a4f0fb7771a932969f3806e/generate.py#L148). `optional_color` isn't a bad name though.

**Comment by dacid44** [comment]
> This PR should also ideally remove all the python code and update the README to give an example on how to override the defaults that we've set to different values.

That's a pretty big change. I can do that, though.

**Comment by sgoudham** [comment]
> That's a pretty big change. I can do that, though.

Sorry, I know that increases the scope of this PR by quite a bit. I'm happy to provide guidance where needed.
I'd say that this PR needs the following changes before it's ready for review:

- [ ] Converts the existing python code into equivalent whiskers template(s)
- [ ] Deletes the following files: `release.sh`, `generate.py`, `requirements.txt` & `cosmic-term.py`
- [ ] Adds a `justfile` at the root of the repository following conventions in the organisation seen elsewhere. (e.g. [catppuccin/foliate](https://github.com/catppuccin/foliate/blob/main/justfile).
- [ ] Updates the README to mention whiskers instead of python and provides sample commands to override the frontmatter variables to tweak the generation of the theme. There is a simple example on the [catppuccin/whiskers README](https://github.com/catppuccin/whiskers?tab=readme-ov-file#overrides) but essentially the `--overrides` flag needs to be present with a JSON string to override the frontmatter.

Hope that's okay @dacid44!

**Comment by dacid44** [comment]
For the main cosmic theme, I was looking through libcosmic, and it looks like they actually don't have palette colors for blue, red, green, or yellow. Did the cosmic palette originally have those colors? Should I match the current cosmic theme layout, or the python file output? https://pop-os.github.io/libcosmic/cosmic/cosmic_theme/struct.CosmicPaletteInner.html

**Comment by dacid44** [comment]
> I understand where you're coming from but I think this sets a bit of a weird precedent for the organisation where we allow our entire brand to be overridable. I'm all for customisation and it's obviously designed into whiskers with the way our colour overrides flag / frontmatter works but it feels odd to have that be overridable by default within the Catppuccin organisation.

@sgoudham Would it be OK to leave the name field in but not document it, or should I take it out completely? I imagine a more common use of it might be for a single user to have a theme which only overrides i.e. the roundness, and name it `CatppuccinSquare` or something just to be distinguishable from the main theme.

**Comment by dacid44** [comment]
I'm marking this as ready to merge, as I think it would be fine if merged now, but there's a couple of outstanding questions:
- Should the template file have an overrideable name (see my last comment)? Currently the name is overrideable but the ability to do so isn't documented in the README.
- Should the `cosmic-settings` palette match the current output of the `generate.py` (which also appears to match how every other COSMIC color scheme I've found works), or should it match the [struct definition from `libcosmic`](https://pop-os.github.io/libcosmic/cosmic/cosmic_theme/struct.CosmicPaletteInner.html) (and the layout that you get if you export a theme in the current version of `cosmic-settings`)? `cosmic-settings` seems to import either one with no issues, so currently it matches the `generate.py`.

File: templates/cosmic-settings.tera
Comment (sgoudham) [inline templates/cosmic-settings.tera]: > I imagine a more common use of it might be for a single user to have a theme which only overrides i.e. the roundness, and name it CatppuccinSquare or something just to be distinguishable from the main theme.

It looks like what we want to follow is our (now archived) [GTK theme zip naming convention](https://github.com/catppuccin/gtk/releases/tag/v1.0.3) where we had a bunch of customisation options but defaulted to some. If the roundness is common enough to be overidden then we should have it like so:

```suggestion
    filename: "themes/cosmic-settings/catppuccin-{{ flavor.identifier }}-{{ accent }}+{{ roundness }}.ron"
```

File: templates/cosmic-term.tera
Comment (sgoudham) [inline templates/cosmic-term.tera]: See comment above

```suggestion
    filename: "themes/cosmic-term/catppuccin-{{ flavor.identifier }}.ron"
```

File: justfile
Comment (sgoudham) [inline justfile]: I'm not sure why we need the folders to be at the top level? I'd want this to be simplified to:

```suggestion
clean:
  rm -rf themes

build: clean
  whiskers templates/cosmic-settings.tera
  whiskers templates/cosmic-term.tera
```

File: .gitignore
Comment (sgoudham) [inline .gitignore]: I don't think we should ignore this and instead put the output of the templates inside of here, like so:

```
themes/
├─ cosmic-term/
├─ cosmic-settings/
```

**Review by sgoudham — CHANGES_REQUESTED** [review]
A few comments from me, thanks for your work so far!

File: templates/cosmic-settings.tera
Comment (dacid44) [reply templates/cosmic-settings.tera]: That makes sense. AFAIK, the roundness is sort of orthogonal to the palette (which is the main concern of catppuccin), but we should have a way to control it anyway, since there's not a way to set just the colors without also setting the roundness. In that case, I could put it back to being used as the "variant" like it was before, so it's generated in the matrix with the flavor and accent, but that's a *lot* of theme variants to be generating (168, unless we want to reduce the number of generated accent colors.) What do you think?

Comment (sgoudham) [reply templates/cosmic-settings.tera]: I'm not so keen on having it as part of the matrix to put 168 files into source control and I wouldn't want to reduce the number of accent generated files either.

We didn't generate the full matrix for GTK assets so I think it's fine putting `{{ roundness }}` into the filename by default since you mention that you can't set the colours without a roundness. The README has information on how to override the options too which is good.

File: justfile
Comment (dacid44) [reply justfile]: The `| true` on the `rm` command makes sure that the justfile doesn't error if the themes directory didn't already exist. I'll make the other adjustments though.

File: justfile
Comment (sgoudham) [inline justfile]: > The | true on the rm command makes sure that the justfile doesn't error if the themes directory didn't already exist.

`rm -rf` doesn't output a non zero exit code if it tries to remove a directory that doesn't exist, see my screenshot below:

image

I think the inclusion of `| true` is more confusing than just doing `rm -rf` since the outcomes are the exact same.

```suggestion
  rm -rf themes
```

File: templates/cosmic-settings.tera
Comment (sgoudham) [inline templates/cosmic-settings.tera]: I'd appreciate if we could lowercase these generated files since it would match ctp/cursors, ctp/gtk. I'd like for lowercase files to be the standard going forward in the organisation.

Also removes the frontmatter variable as discussed.

```suggestion
roundness: "round"
window_hint_color: ""
bg_alpha: 1.0
frosted: false
outer_gap_size: 0
inner_gap_size: 8
active_hint_size: 3
whiskers:
    version: 2.5.1
    matrix:
    - flavor
    - accent
    filename: "themes/cosmic-settings/catppuccin-{{ flavor.identifier }}-{{ accent }}+{{ roundness }}.ron"
```

File: templates/cosmic-term.tera
Comment (sgoudham) [inline templates/cosmic-term.tera]:
```suggestion
whiskers:
    version: 2.5.1
    matrix:
    - flavor
    filename: "themes/cosmic-term/catppuccin-{{ flavor.identifier }}.ron"
```

**Review by sgoudham — CHANGES_REQUESTED** [review]
A few edits to be made as per our discussion

File: templates/cosmic-settings.tera
Comment (dacid44) [reply templates/cosmic-settings.tera]: Should the names in the theme name (inside the generated files, not the filename) be lowercase or uppercase?

File: justfile
Comment (dacid44) [reply justfile]: Gotcha, I didn't see you'd suggested the `-r` -> `-rf` change.

File: templates/cosmic-settings.tera
Comment (sgoudham) [reply templates/cosmic-settings.tera]: I believe the name inside the generated files is the "pretty" name to be displayed on the user interface so I think it makes sense to keep those the way they are.

I don't use COSMIC so I would need someone to confirm that for me.

Comment (dacid44) [reply templates/cosmic-settings.tera]: AFAIK the settings theme name isn't displayed anywhere in the interface, but given the field exists, it may at some point. The terminal theme is displayed in the interface though.

**Review by sgoudham — APPROVED** [review]
Just updated the README to be a bit more concise, think it's good to merge now!

I'll merge in a few days if @EdenQwQ doesn't get around to giving it a final review.

Thank you again for all your work!

**Pull request merged**
